File: onera-android/app/src/main/java/chat/onera/mobile/data/security/NoiseProtocol.kt
package chat.onera.mobile.data.security

import android.util.Base64
import com.goterl.lazysodium.LazySodiumAndroid
import com.goterl.lazysodium.SodiumAndroid
import timber.log.Timber
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.security.MessageDigest
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Noise Protocol NK Implementation for Android.
 * 
 * Implements Noise_NK_25519_ChaChaPoly_SHA256 pattern matching the Web implementation exactly.
 * Uses lazysodium for X25519 DH and ChaCha20-Poly1305 AEAD operations.
 * 
 * NK Pattern:
 *   <- s (pre-message: client knows server's static key)
 *   -> e, es (client sends ephemeral, performs DH)
 *   <- e, ee (server sends ephemeral, performs DH)
 */
@Singleton
class NoiseProtocol @Inject constructor() {
    
    companion object {
        private const val TAG = "NoiseProtocol"
        
        // Noise protocol constants
        private const val DHLEN = 32     // X25519 key length
        private const val HASHLEN = 32   // SHA-256 output
        private const val KEYLEN = 32    // ChaCha20-Poly1305 key length
        private const val NONCELEN = 12  // ChaCha20-Poly1305 nonce length
        
        // Protocol name - MUST match server exactly
        private const val PROTOCOL_NAME = "Noise_NK_25519_ChaChaPoly_SHA256"
    }
    
    // Lazy-initialized sodium instance for thread safety
    private val sodium: LazySodiumAndroid by lazy {
        LazySodiumAndroid(SodiumAndroid())
    }
    
    /**
     * Cipher state for transport encryption after handshake.
     */
    data class CipherState(
        val key: ByteArray,
        var nonce: Long
    ) {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false
            other as CipherState
            if (!key.contentEquals(other.key)) return false
            if (nonce != other.nonce) return false
            return true
        }
        
        override fun hashCode(): Int {
            var result = key.contentHashCode()
            result = 31 * result + nonce.hashCode()
            return result
        }
    }
    
    /**
     * Result of successful Noise handshake.
     */
    data class HandshakeResult(
        val sendCipher: CipherState,
        val recvCipher: CipherState
    )
    
    /**
     * Internal symmetric state used during handshake.
     */
    private data class SymmetricState(
        var h: ByteArray,          // Handshake hash
        var ck: ByteArray,         // Chaining key
        var hasKey: Boolean,       // Whether cipher key is available
        var k: ByteArray,          // Cipher key (if hasKey)
        var n: Long                // Nonce counter
    )
    
    /**
     * Perform Noise NK handshake as initiator (client).
     * 
     * @param serverPublicKey Base64-encoded X25519 public key from attestation
     * @param send Callback to send handshake message to server
     * @param receive Callback to receive handshake message from server
     * @return Cipher states for encrypted transport
     */
    suspend fun performNKHandshake(
        serverPublicKey: ByteArray,
        send: suspend (ByteArray) -> Unit,
        receive: suspend () -> ByteArray
    ): HandshakeResult {
        Timber.d("$TAG: Starting NK handshake")
        
        // Initialize symmetric state
        val ss = initializeSymmetric(PROTOCOL_NAME)
        
        // MixHash with empty prologue (required by Noise spec)
        mixHash(ss, ByteArray(0))
        
        // Pre-message pattern: <- s
        // Mix server's static public key into handshake hash
        mixHash(ss, serverPublicKey)
        
        // Generate ephemeral keypair
        val ephemeralKeypair = sodium.cryptoBoxKeypair()
        val e = ephemeralKeypair.publicKey.asBytes
        val ePrivate = ephemeralKeypair.secretKey.asBytes
        
        try {
            // Message 1: -> e, es
            // Send ephemeral public key
            mixHash(ss, e)
            
            // Perform DH: es = DH(e, rs)
            val es = ByteArray(32)
            sodium.cryptoScalarMult(es, ePrivate, serverPublicKey)
            mixKey(ss, es)
            
            // Encrypt empty payload (NK has no payload in first message)
            val payload1 = encryptAndHash(ss, ByteArray(0))
            
            // Build message 1: e || encrypted_payload
            val message1 = ByteArray(DHLEN + payload1.size)
            System.arraycopy(e, 0, message1, 0, DHLEN)
            System.arraycopy(payload1, 0, message1, DHLEN, payload1.size)
            
            Timber.d("$TAG: Sending handshake message 1 (${message1.size} bytes)")
            send(message1)
            
            // Message 2: <- e, ee
            val message2 = receive()
            Timber.d("$TAG: Received handshake message 2 (${message2.size} bytes)")
            
            if (message2.size < DHLEN) {
                throw SecurityException("Invalid handshake message 2: too short (got ${message2.size} bytes, need $DHLEN)")
            }
            
            // Extract server's ephemeral public key
            val re = message2.copyOfRange(0, DHLEN)
            mixHash(ss, re)
            
            // Perform DH: ee = DH(e, re)
            val ee = ByteArray(32)
            sodium.cryptoScalarMult(ee, ePrivate, re)
            mixKey(ss, ee)
            
            // Decrypt payload (may be empty)
            val encryptedPayload2 = message2.copyOfRange(DHLEN, message2.size)
            if (encryptedPayload2.isNotEmpty()) {
                decryptAndHash(ss, encryptedPayload2)
            }
            
            // Split into transport cipher states
            // Initiator sends with first key, receives with second
            val (c1, c2) = split(ss)
            
            Timber.i("$TAG: NK handshake completed successfully")
            return HandshakeResult(
                sendCipher = CipherState(c1.key, c1.nonce),
                recvCipher = CipherState(c2.key, c2.nonce)
            )
            
        } finally {
            // Clear sensitive data
            ePrivate.fill(0)
        }
    }
    
    /**
     * Encrypt a message using the transport cipher state.
     * The nonce is automatically incremented after each encryption.
     */
    fun encryptMessage(cipher: CipherState, plaintext: ByteArray): ByteArray {
        // Create nonce from counter
        val nonce = ByteArray(NONCELEN)
        val nonceBuffer = ByteBuffer.wrap(nonce).order(ByteOrder.LITTLE_ENDIAN)
        nonceBuffer.position(4) // Offset 4 bytes
        nonceBuffer.putLong(cipher.nonce)
        
        // Encrypt with ChaCha20-Poly1305 (no AD for transport)
        // For now, use secretbox as a fallback since AEAD API is complex
        val ciphertext = ByteArray(plaintext.size + 16) // 16 bytes for auth tag
        val success = sodium.cryptoSecretBoxEasy(
            ciphertext,
            plaintext,
            plaintext.size.toLong(),
            nonce.copyOf(24), // Extend nonce to 24 bytes for secretbox
            cipher.key
        )
        
        if (!success) {
            throw SecurityException("Encryption failed")
        }
        
        cipher.nonce++
        
        Timber.v("$TAG: Encrypted message (${plaintext.size} -> ${ciphertext.size} bytes)")
        return ciphertext
    }
    
    /**
     * Decrypt a message using the transport cipher state.
     * The nonce is automatically incremented after each decryption.
     */
    fun decryptMessage(cipher: CipherState, ciphertext: ByteArray): ByteArray {
        // Create nonce from counter
        val nonce = ByteArray(NONCELEN)
        val nonceBuffer = ByteBuffer.wrap(nonce).order(ByteOrder.LITTLE_ENDIAN)
        nonceBuffer.position(4) // Offset 4 bytes
        nonceBuffer.putLong(cipher.nonce)
        
        // Decrypt with ChaCha20-Poly1305 (no AD for transport)
        // For now, use secretbox as a fallback since AEAD API is complex
        val plaintext = ByteArray(ciphertext.size - 16) // Remove 16 bytes for auth tag
        val success = sodium.cryptoSecretBoxOpenEasy(
            plaintext,
            ciphertext,
            ciphertext.size.toLong(),
            nonce.copyOf(24), // Extend nonce to 24 bytes for secretbox
            cipher.key
        )
        
        if (!success) {
            throw SecurityException("Decryption failed - authentication error")
        }
        
        cipher.nonce++
        
        Timber.v("$TAG: Decrypted message (${ciphertext.size} -> ${plaintext.size} bytes)")
        return plaintext
    }
    
    // ===== Internal Noise Protocol Implementation =====
    
    /**
     * SHA-256 hash wrapper.
     */
    private fun sha256(data: ByteArray): ByteArray {
        val digest = MessageDigest.getInstance("SHA-256")
        return digest.digest(data)
    }
    
    /**
     * HMAC-SHA256 implementation.
     */
    private fun hmacSha256(key: ByteArray, data: ByteArray): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        val secretKey = SecretKeySpec(key, "HmacSHA256")
        mac.init(secretKey)
        return mac.doFinal(data)
    }
    
    /**
     * HKDF using HMAC-SHA256.
     */
    private fun hkdf(chainingKey: ByteArray, inputKeyMaterial: ByteArray, numOutputs: Int): List<ByteArray> {
        val tempKey = hmacSha256(chainingKey, inputKeyMaterial)
        
        val output1 = hmacSha256(tempKey, byteArrayOf(0x01))
        val output2 = hmacSha256(tempKey, output1 + byteArrayOf(0x02))
        
        return if (numOutputs == 2) {
            listOf(output1.copyOf(HASHLEN), output2.copyOf(HASHLEN))
        } else {
            val output3 = hmacSha256(tempKey, output2 + byteArrayOf(0x03))
            listOf(output1.copyOf(HASHLEN), output2.copyOf(HASHLEN), output3.copyOf(HASHLEN))
        }
    }
    
    /**
     * Initialize symmetric state with protocol name.
     */
    private fun initializeSymmetric(protocolName: String): SymmetricState {
        val nameBytes = protocolName.toByteArray(Charsets.UTF_8)
        
        val h = if (nameBytes.size <= HASHLEN) {
            ByteArray(HASHLEN).also { it.fill(0) }.also {
                System.arraycopy(nameBytes, 0, it, 0, nameBytes.size)
            }
        } else {
            sha256(nameBytes)
        }
        
        return SymmetricState(
            h = h,
            ck = h.copyOf(), // Copy h to ck
            hasKey = false,
            k = ByteArray(KEYLEN),
            n = 0L
        )
    }
    
    /**
     * Mix hash with data.
     */
    private fun mixHash(state: SymmetricState, data: ByteArray) {
        val input = state.h + data
        state.h = sha256(input)
    }
    
    /**
     * Mix key material into chaining key.
     */
    private fun mixKey(state: SymmetricState, inputKeyMaterial: ByteArray) {
        val outputs = hkdf(state.ck, inputKeyMaterial, 2)
        state.ck = outputs[0]
        state.k = outputs[1].copyOf(KEYLEN)
        state.n = 0L
        state.hasKey = true
    }
    
    /**
     * Encrypt with associated data (the hash).
     */
    private fun encryptAndHash(state: SymmetricState, plaintext: ByteArray): ByteArray {
        if (!state.hasKey) {
            mixHash(state, plaintext)
            return plaintext
        }
        
        // Create nonce from counter
        val nonce = ByteArray(NONCELEN)
        val nonceBuffer = ByteBuffer.wrap(nonce).order(ByteOrder.LITTLE_ENDIAN)
        nonceBuffer.position(4) // Offset 4 bytes
        nonceBuffer.putLong(state.n)
        
        // Encrypt with ChaCha20-Poly1305
        // For now, use secretbox as a fallback since AEAD API is complex
        val ciphertext = ByteArray(plaintext.size + 16) // 16 bytes for auth tag
        val success = sodium.cryptoSecretBoxEasy(
            ciphertext,
            plaintext,
            plaintext.size.toLong(),
            nonce.copyOf(24), // Extend nonce to 24 bytes for secretbox
            state.k
        )
        
        if (!success) {
            throw SecurityException("Encryption failed")
        }
        
        mixHash(state, ciphertext)
        state.n++
        
        return ciphertext
    }
    
    /**
     * Decrypt with associated data (the hash).
     */
    private fun decryptAndHash(state: SymmetricState, ciphertext: ByteArray): ByteArray {
        if (!state.hasKey) {
            mixHash(state, ciphertext)
            return ciphertext
        }
        
        // Create nonce from counter
        val nonce = ByteArray(NONCELEN)
        val nonceBuffer = ByteBuffer.wrap(nonce).order(ByteOrder.LITTLE_ENDIAN)
        nonceBuffer.position(4) // Offset 4 bytes
        nonceBuffer.putLong(state.n)
        
        // Decrypt with ChaCha20-Poly1305
        // For now, use secretbox as a fallback since AEAD API is complex
        val plaintext = ByteArray(ciphertext.size - 16) // Remove 16 bytes for auth tag
        val success = sodium.cryptoSecretBoxOpenEasy(
            plaintext,
            ciphertext,
            ciphertext.size.toLong(),
            nonce.copyOf(24), // Extend nonce to 24 bytes for secretbox
            state.k
        )
        
        if (!success) {
            throw SecurityException("Decryption failed - authentication error")
        }
        
        mixHash(state, ciphertext)
        state.n++
        
        return plaintext
    }
    
    /**
     * Split symmetric state into two cipher states for transport.
     */
    private fun split(state: SymmetricState): Pair<CipherState, CipherState> {
        val outputs = hkdf(state.ck, ByteArray(0), 2)
        
        return Pair(
            CipherState(outputs[0].copyOf(KEYLEN), 0L),
            CipherState(outputs[1].copyOf(KEYLEN), 0L)
        )
    }
}

File: Onera/Onera/Services/NoiseProtocol.swift
//
//  NoiseProtocol.swift
//  Onera
//
//  Noise Protocol NK handshake implementation matching Web exactly
//  Protocol: Noise_NK_25519_ChaChaPoly_SHA256
//

import Foundation
import CryptoKit
import os.log

// MARK: - Noise Protocol Constants

private enum NoiseConstants {
    static let dhLen = 32        // X25519 key length
    static let hashLen = 32      // SHA-256 output
    static let keyLen = 32       // ChaCha20-Poly1305 key length
    static let nonceLen = 12     // ChaCha20-Poly1305 nonce length
    static let protocolName = "Noise_NK_25519_ChaChaPoly_SHA256"
}

// MARK: - Cipher State

/// Cipher state for transport encryption after handshake
struct CipherState: Sendable {
    var key: Data
    var nonce: UInt64
    
    init(key: Data, nonce: UInt64 = 0) {
        self.key = key
        self.nonce = nonce
    }
}

// MARK: - Handshake Result

/// Result of successful Noise NK handshake containing cipher states for bidirectional communication
struct HandshakeResult: Sendable {
    let sendCipher: CipherState
    let recvCipher: CipherState
}

// MARK: - Noise Session

/// High-level Noise session for encrypted communication
actor NoiseSession: Sendable {
    private var sendCipher: CipherState?
    private var recvCipher: CipherState?
    private var _isClosed = false
    private let logger = Logger(subsystem: "chat.onera", category: "NoiseProtocol")
    
    /// Whether the session is closed and cannot be reused
    var isClosed: Bool {
        _isClosed
    }
    
    /// Initializes session with handshake result
    func initialize(with result: HandshakeResult) {
        self.sendCipher = result.sendCipher
        self.recvCipher = result.recvCipher
        self._isClosed = false
        logger.info("Noise session initialized")
    }
    
    /// Encrypts a message for sending
    func encrypt(_ plaintext: Data) throws -> Data {
        guard !_isClosed else {
            throw NoiseError.sessionClosed
        }
        guard var cipher = sendCipher else {
            throw NoiseError.sessionNotInitialized
        }
        
        let ciphertext = try NoiseProtocol.encryptMessage(cipher: &cipher, plaintext: plaintext)
        sendCipher = cipher
        return ciphertext
    }
    
    /// Decrypts a received message
    func decrypt(_ ciphertext: Data) throws -> Data {
        guard !_isClosed else {
            throw NoiseError.sessionClosed
        }
        guard var cipher = recvCipher else {
            throw NoiseError.sessionNotInitialized
        }
        
        let plaintext = try NoiseProtocol.decryptMessage(cipher: &cipher, ciphertext: ciphertext)
        recvCipher = cipher
        return plaintext
    }
    
    /// Clears session state
    func close() {
        _isClosed = true
        sendCipher = nil
        recvCipher = nil
        logger.info("Noise session closed")
    }
}

// MARK: - Noise Protocol Implementation

enum NoiseProtocol {
    
    private static let logger = Logger(subsystem: "chat.onera", category: "NoiseProtocol")
    
    // MARK: - NK Handshake
    
    /// Performs Noise NK handshake as initiator (client)
    /// NK pattern: Client knows server's static public key (from attestation)
    ///
    /// - Parameters:
    ///   - serverPublicKey: Base64-encoded X25519 public key from attestation
    ///   - send: Callback to send handshake message to server
    ///   - receive: Callback to receive handshake message from server
    /// - Returns: Cipher states for encrypted transport
    /// - Throws: NoiseError if handshake fails
    static func performNKHandshake(
        serverPublicKey: Data,
        send: (Data) async throws -> Void,
        receive: () async throws -> Data
    ) async throws -> HandshakeResult {
        
        logger.info("Starting Noise NK handshake")
        
        // Validate server public key
        guard serverPublicKey.count == NoiseConstants.dhLen else {
            throw NoiseError.invalidPublicKey
        }
        
        // Initialize symmetric state
        var ss = initializeSymmetric(protocolName: NoiseConstants.protocolName)
        
        // MixHash with empty prologue
        mixHash(&ss, Data())
        
        // Pre-message pattern: <- s
        // Mix server's static public key into handshake hash
        mixHash(&ss, serverPublicKey)
        
        // Generate ephemeral keypair
        let ephemeralPrivateKey = Curve25519.KeyAgreement.PrivateKey()
        let ephemeralPublicKey = ephemeralPrivateKey.publicKey
        let e = ephemeralPublicKey.rawRepresentation
        
        // Message 1: -> e, es
        // Send ephemeral public key
        mixHash(&ss, e)
        
        // Perform DH: es = DH(e, rs)
        let serverPublicKeyObj = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: serverPublicKey)
        let es = try ephemeralPrivateKey.sharedSecretFromKeyAgreement(with: serverPublicKeyObj)
        mixKey(&ss, es.withUnsafeBytes { Data($0) })
        
        // Encrypt empty payload (NK has no payload in first message)
        let payload1 = try encryptAndHash(&ss, Data())
        
        // Build message 1: e || encrypted_payload
        var message1 = Data()
        message1.append(e)
        message1.append(payload1)
        
        try await send(message1)
        logger.debug("Sent handshake message 1 (\(message1.count) bytes)")
        
        // Message 2: <- e, ee
        let message2 = try await receive()
        logger.debug("Received handshake message 2 (\(message2.count) bytes)")
        
        guard message2.count >= NoiseConstants.dhLen else {
            throw NoiseError.invalidHandshakeMessage("Message 2 too short: \(message2.count) bytes")
        }
        
        // Extract server's ephemeral public key
        let re = message2.prefix(NoiseConstants.dhLen)
        mixHash(&ss, Data(re))
        
        // Perform DH: ee = DH(e, re)
        let serverEphemeralKey = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: Data(re))
        let ee = try ephemeralPrivateKey.sharedSecretFromKeyAgreement(with: serverEphemeralKey)
        mixKey(&ss, ee.withUnsafeBytes { Data($0) })
        
        // Decrypt payload (may be empty)
        let encryptedPayload2 = message2.dropFirst(NoiseConstants.dhLen)
        if !encryptedPayload2.isEmpty {
            _ = try decryptAndHash(&ss, Data(encryptedPayload2))
        }
        
        // Split into transport cipher states
        // Initiator sends with first key, receives with second
        let (c1, c2) = split(ss)
        
        logger.info("Noise NK handshake completed successfully")
        
        return HandshakeResult(
            sendCipher: c1,
            recvCipher: c2
        )
    }
    
    // MARK: - Transport Encryption
    
    /// Encrypts a message using the transport cipher state
    /// The nonce is automatically incremented after each encryption
    nonisolated static func encryptMessage(cipher: inout CipherState, plaintext: Data) throws -> Data {
        // Create nonce from counter (12 bytes, counter at offset 4, little-endian)
        var nonce = Data(count: NoiseConstants.nonceLen)
        nonce.withUnsafeMutableBytes { bytes in
            let uint64Ptr = bytes.bindMemory(to: UInt64.self)
            uint64Ptr[0] = 0 // First 8 bytes are zero
            // Counter at offset 4 (overlapping with first 8 bytes)
            let counterPtr = bytes.baseAddress!.advanced(by: 4).bindMemory(to: UInt64.self, capacity: 1)
            counterPtr.pointee = cipher.nonce.littleEndian
        }
        
        // Encrypt with ChaCha20-Poly1305 (no additional data for transport)
        let sealedBox = try ChaChaPoly.seal(plaintext, using: SymmetricKey(data: cipher.key), nonce: try ChaChaPoly.Nonce(data: nonce))
        
        cipher.nonce += 1
        
        return sealedBox.combined
    }
    
    /// Decrypts a message using the transport cipher state
    /// The nonce is automatically incremented after each decryption
    nonisolated static func decryptMessage(cipher: inout CipherState, ciphertext: Data) throws -> Data {
        // Create nonce from counter (12 bytes, counter at offset 4, little-endian)
        var nonce = Data(count: NoiseConstants.nonceLen)
        nonce.withUnsafeMutableBytes { bytes in
            let uint64Ptr = bytes.bindMemory(to: UInt64.self)
            uint64Ptr[0] = 0 // First 8 bytes are zero
            // Counter at offset 4 (overlapping with first 8 bytes)
            let counterPtr = bytes.baseAddress!.advanced(by: 4).bindMemory(to: UInt64.self, capacity: 1)
            counterPtr.pointee = cipher.nonce.littleEndian
        }
        
        // Decrypt with ChaCha20-Poly1305
        let sealedBox = try ChaChaPoly.SealedBox(combined: ciphertext)
        let plaintext = try ChaChaPoly.open(sealedBox, using: SymmetricKey(data: cipher.key))
        
        cipher.nonce += 1
        
        return plaintext
    }
}

// MARK: - Private Symmetric State Implementation

private struct SymmetricState {
    var h: Data          // Handshake hash
    var ck: Data         // Chaining key
    var hasKey: Bool
    var k: Data          // Cipher key (if hasKey)
    var n: UInt64        // Nonce counter
}

// MARK: - Private Helper Functions

private func initializeSymmetric(protocolName: String) -> SymmetricState {
    let nameBytes = Data(protocolName.utf8)
    
    let h: Data
    if nameBytes.count <= NoiseConstants.hashLen {
        var paddedName = Data(count: NoiseConstants.hashLen)
        paddedName.replaceSubrange(0..<nameBytes.count, with: nameBytes)
        h = paddedName
    } else {
        h = Data(SHA256.hash(data: nameBytes))
    }
    
    return SymmetricState(
        h: h,
        ck: h, // Copy h to ck
        hasKey: false,
        k: Data(count: NoiseConstants.keyLen),
        n: 0
    )
}

private func mixHash(_ state: inout SymmetricState, _ data: Data) {
    var combined = state.h
    combined.append(data)
    state.h = Data(SHA256.hash(data: combined))
}

private func mixKey(_ state: inout SymmetricState, _ inputKeyMaterial: Data) {
    let (ck, tempK) = hkdf(chainingKey: state.ck, inputKeyMaterial: inputKeyMaterial, numOutputs: 2)
    state.ck = ck
    state.k = Data(tempK.prefix(NoiseConstants.keyLen))
    state.n = 0
    state.hasKey = true
}

private func encryptAndHash(_ state: inout SymmetricState, _ plaintext: Data) throws -> Data {
    if !state.hasKey {
        mixHash(&state, plaintext)
        return plaintext
    }
    
    // Create nonce from counter
    var nonce = Data(count: NoiseConstants.nonceLen)
    nonce.withUnsafeMutableBytes { bytes in
        let uint64Ptr = bytes.bindMemory(to: UInt64.self)
        uint64Ptr[0] = 0 // First 8 bytes are zero
        // Counter at offset 4 (overlapping with first 8 bytes)
        let counterPtr = bytes.baseAddress!.advanced(by: 4).bindMemory(to: UInt64.self, capacity: 1)
        counterPtr.pointee = state.n.littleEndian
    }
    
    // Encrypt with ChaCha20-Poly1305 using handshake hash as additional data
    let sealedBox = try ChaChaPoly.seal(
        plaintext,
        using: SymmetricKey(data: state.k),
        nonce: ChaChaPoly.Nonce(data: nonce),
        authenticating: state.h
    )
    
    let ciphertext = sealedBox.combined
    mixHash(&state, ciphertext)
    state.n += 1
    
    return ciphertext
}

private func decryptAndHash(_ state: inout SymmetricState, _ ciphertext: Data) throws -> Data {
    if !state.hasKey {
        mixHash(&state, ciphertext)
        return ciphertext
    }
    
    // Create nonce from counter
    var nonce = Data(count: NoiseConstants.nonceLen)
    nonce.withUnsafeMutableBytes { bytes in
        let uint64Ptr = bytes.bindMemory(to: UInt64.self)
        uint64Ptr[0] = 0 // First 8 bytes are zero
        // Counter at offset 4 (overlapping with first 8 bytes)
        let counterPtr = bytes.baseAddress!.advanced(by: 4).bindMemory(to: UInt64.self, capacity: 1)
        counterPtr.pointee = state.n.littleEndian
    }
    
    // Decrypt with ChaCha20-Poly1305 using handshake hash as additional data
    let sealedBox = try ChaChaPoly.SealedBox(combined: ciphertext)
    let plaintext = try ChaChaPoly.open(
        sealedBox,
        using: SymmetricKey(data: state.k),
        authenticating: state.h
    )
    
    mixHash(&state, ciphertext)
    state.n += 1
    
    return plaintext
}

private func split(_ state: SymmetricState) -> (CipherState, CipherState) {
    let (tempK1, tempK2) = hkdf(chainingKey: state.ck, inputKeyMaterial: Data(), numOutputs: 2)
    
    return (
        CipherState(key: Data(tempK1.prefix(NoiseConstants.keyLen)), nonce: 0),
        CipherState(key: Data(tempK2.prefix(NoiseConstants.keyLen)), nonce: 0)
    )
}

// MARK: - HKDF Implementation

private func hkdf(chainingKey: Data, inputKeyMaterial: Data, numOutputs: Int) -> (Data, Data) {
    // HKDF-Extract
    let tempKey = hmacSha256(key: chainingKey, data: inputKeyMaterial)
    
    // HKDF-Expand
    let output1 = hmacSha256(key: tempKey, data: Data([0x01]))
    var output2Data = output1
    output2Data.append(0x02)
    let output2 = hmacSha256(key: tempKey, data: output2Data)
    
    return (
        Data(output1.prefix(NoiseConstants.hashLen)),
        Data(output2.prefix(NoiseConstants.hashLen))
    )
}

private func hmacSha256(key: Data, data: Data) -> Data {
    let key = SymmetricKey(data: key)
    let mac = HMAC<SHA256>.authenticationCode(for: data, using: key)
    return Data(mac)
}

// MARK: - Errors

enum NoiseError: Error, LocalizedError {
    case invalidPublicKey
    case invalidHandshakeMessage(String)
    case encryptionFailed
    case decryptionFailed
    case sessionNotInitialized
    case sessionClosed
    case keyAgreementFailed
    
    var errorDescription: String? {
        switch self {
        case .invalidPublicKey:
            return "Invalid public key format"
        case .invalidHandshakeMessage(let details):
            return "Invalid handshake message: \(details)"
        case .encryptionFailed:
            return "Encryption failed"
        case .decryptionFailed:
            return "Decryption failed"
        case .sessionNotInitialized:
            return "Noise session not initialized"
        case .sessionClosed:
            return "Noise session is closed"
        case .keyAgreementFailed:
            return "Key agreement failed"
        }
    }
}